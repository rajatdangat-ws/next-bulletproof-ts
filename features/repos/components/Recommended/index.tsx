/**
 *
 * Recommended
 *
 */

import React from "react";
import { Row, Col } from "antd";
import { useRouter } from "next/router";
import { ClickableTags } from "@common";
import { Recommendation } from "@features/repos/types";

interface RecommendedProps {
  recommendations?: Recommendation[];
}

const Recommended: React.FC<RecommendedProps> = ({ recommendations }) => {
  const router = useRouter();

  return (
    <Row data-testid="recommended">
      {recommendations?.length &&
        recommendations.map(({ id, name }) => (
          <Col key={id} onClick={() => router.push(`/info/${name}`)}>
            <ClickableTags>{name}</ClickableTags>
          </Col>
        ))}
    </Row>
  );
};

export default Recommended;
