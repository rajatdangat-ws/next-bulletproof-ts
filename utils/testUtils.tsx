import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { IntlProvider } from "react-intl";
import { Provider as ReduxProvider } from "react-redux";
import { store } from "../store";
import messages from "../translations/en.json";

/**
 * Renders the passed in components or tree with all the providers.
 * To configure just add more providers and use the function in tests to render with providers
 * @param children, all the components in the composition
 * @returns React.FC, renders the DOM with all the providers
 */
export const WithAllProviders: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <IntlProvider messages={messages} locale="en" defaultLocale="en">
        {children}
      </IntlProvider>
    </ReduxProvider>
  );
};

/**
 * Generic render function that renders all the listed components with all the providers
 * @param ui The ReactElement (ie. Component) to render
 * @param options extra render options
 * @returns returns the RenderResult with all the things that react-testing library provides
 */
export const renderForTesting = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: WithAllProviders, ...options });

/**
 * Stops execution in tests for a specified milliseconds
 * @param ms time in milliseconds, how long to stop
 * @returns returns a new Promise to hang
 */
export const timeout = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Generate an API Response that can be used to mock API Responses
 * @param ok boolean specifing will the response resolve or reject
 * @param data the data that is to be sent back
 * @returns APIResponse consisting ok and data
 */
export function apiResponseGenerator<T>(ok: boolean, data: T) {
  return {
    ok,
    data,
  };
}

export * from "@testing-library/react";
export { renderForTesting as render };
