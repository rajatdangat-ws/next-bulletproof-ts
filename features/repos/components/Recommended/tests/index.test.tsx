/**
 *
 * Tests for Recommended
 *
 */

import React from "react";
import { render } from "@utils/testUtils";
import Recommended from "../index";

describe("<Recommended />", () => {
  const props = {
    recommendations: [
      {
        name: "test repo name",
        id: 1,
      },
    ],
  };

  it("should render and match the snapshot", () => {
    const { baseElement } = render(<Recommended {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should contain 1 Recommended component", () => {
    const { getAllByTestId } = render(<Recommended {...props} />);
    expect(getAllByTestId("recommended").length).toBe(1);
  });
});
