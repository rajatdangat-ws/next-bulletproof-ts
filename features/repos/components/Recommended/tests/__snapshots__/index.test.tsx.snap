// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recommended /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-row"
      data-testid="recommended"
    >
      <div
        class="ant-col"
      >
        <span
          class="ant-tag styled__ClickableTags-sc-15gkald-2 kHFdpo"
        >
          test repo name
        </span>
      </div>
    </div>
  </div>
</body>
`;
